src/table/merge_iterator.rs

```diff
@@ -0,0 +1,362 @@
+use bytes::{Bytes, BytesMut};
+use enum_dispatch::enum_dispatch;
+use std::sync::Arc;
+
+use super::iterator::Iterator;
+use super::TableInner;
+use crate::structs::AgateIterator;
+use crate::util::{KeyComparator, COMPARATOR};
+use crate::Value;
+
+type TableIterator = Iterator<Arc<TableInner>>;
+
+/// `Iterators` includes all iterator types for AgateDB.
+/// By packing them into an enum, we could reduce the
+/// overhead of dynamic dispatch.
+#[enum_dispatch(AgateIterator)]
+pub enum Iterators {
+    MergeIterator(MergeIterator),
```

Comment:
You can mark it `Box` to get rid of the `Box` at L40.

Reply:
I tried using something like this:

```
#[enum_dispatch(Box<AgateIterator>)]
pub enum Iterators {
    MergeIterator(Box<MergeIterator>),
    TableIterator(Box<TableIterator>),
    #[cfg(test)]
    VecIterator(Box<tests::VecIterator>),
}
```

But this breaks method delegation (e.g. `iter.key()` won't work) and breaks `Iterators::from(TableIterator)`. Let me know if I misunderstood something.

Reply:
What I'm expected is
```
pub enum Iterators {
    MergeIterator(Box<MergeIterator>),
    TableIterator(TableIterator),
    #[cfg(test)]
    VecIterator(Box<tests::VecIterator>),
}
```.

On the other hand, is It possible to define `MergeIterator` as
```
struct IteratorNode<I> {
    valid: bool,
    key: BytesMut,
    iter: I,
}

pub struct MergeIterator<L, R> {
    left: IteratorNode<L>,
    right: IteratorNode<R>,
    is_left_small: bool,
    reverse: bool,
    current_key: BytesMut,
}
```?

Reply:
For the former one, I'll have a try. It seems that Box<MergeIterator> doesn't implement `AgateIterator`, not sure if this would work.

For the latter one, that's exactly what I've tried before. There are several issues with this.
* First, we cannot determines type of iterator at compile-time. Our reference implementation would construct one iterator recursively from multiple iterators, when iterating multiple SSTs. It's impossible to know how many SSTs to iterator at compile time.
* Using such implementation make the merge logic very complex (as left and right are treated as different types).

Reply:
From what I can see, the iterator is always be constructed from a slice of iterator? Why not store them as slice in place? So we don't need allocations and probably get better performance as it's the same direction as prefetch.

Reply:
Yes, I agree that iterators can be always stored and used as slice. But this makes the merge iterating logic very complex. We may have to build a heap to store the elements from each iterator before serving them to developers. I suggest trying this later and just keep it the same as reference implementation, by converting a slice of iterators into a binary tree of merge iterators.